### ISSUE TITLE
AWS i3.8xlarge - Under heavy read load NIC gets Tx timeout and gets reset, loses smp_affinity configurations and all IRQs are handled by cpu0 only thus limiting throughput.

### ISSUE THREAD

**Issue by gnumoreno** [scylladb/scylla#3408]
*Installation details*
Scylla version (or git commit hash): Scylla 2.1.3 AMI
Cluster size: 1 node
OS (RHEL/CentOS/Ubuntu/AWS AMI):  us-west-2 | ami-c6e793be
CentOS Linux release 7.2.1511 (Core)

*Hardware details (for performance issues)
Platform (physical/VM/cloud instance type/docker): AWS _i3.8xlarge_
Hardware: 32vcpus | 244GB RAM | 10 Gigabit NIC 
Disks: (SSD/HDD, count): 4 x 1.9 NVMe SSD

NIC Driver: 
```
ethtool -i eth0
driver: ena
version: 1.4.0U
firmware-version: 
bus-info: 0000:00:03.0
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: no
```

perftune.yaml
```
cat /etc/scylla.d/perftune.yaml
cpu_mask: '0xffffffff'
mode: sq_split
nic: eth0
tune:
```

perftune output
```
sudo ./perftune.py --options-file /etc/scylla.d/perftune.yaml
irqbalance is not running
Setting a physical interface eth0...
Setting mask 00000001 in /proc/irq/248/smp_affinity
Setting mask 00000001 in /proc/irq/249/smp_affinity
Setting mask 00000001 in /proc/irq/247/smp_affinity
Setting mask 00000001 in /proc/irq/243/smp_affinity
Setting mask 00010000 in /proc/irq/244/smp_affinity
Setting mask 00010000 in /proc/irq/246/smp_affinity
Setting mask 00010000 in /proc/irq/250/smp_affinity
Setting mask 00010000 in /proc/irq/245/smp_affinity
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-7/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-5/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-3/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-1/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-6/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-4/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-2/rps_cpus
Setting mask fffefffe in /sys/class/net/eth0/queues/rx-0/rps_cpus
Setting net.core.rps_sock_flow_entries to 32768
Setting limit 4096 in /sys/class/net/eth0/queues/rx-7/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-5/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-3/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-1/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-6/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-4/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-2/rps_flow_cnt
Setting limit 4096 in /sys/class/net/eth0/queues/rx-0/rps_flow_cnt
Trying to enable ntuple filtering HW offload for eth0...not supported
Setting mask 00030003 in /sys/class/net/eth0/queues/tx-6/xps_cpus
Setting mask 000c000c in /sys/class/net/eth0/queues/tx-4/xps_cpus
Setting mask 00300030 in /sys/class/net/eth0/queues/tx-2/xps_cpus
Setting mask 00c000c0 in /sys/class/net/eth0/queues/tx-0/xps_cpus
Setting mask 03000300 in /sys/class/net/eth0/queues/tx-7/xps_cpus
Setting mask 0c000c00 in /sys/class/net/eth0/queues/tx-5/xps_cpus
Setting mask 30003000 in /sys/class/net/eth0/queues/tx-3/xps_cpus
Setting mask c000c000 in /sys/class/net/eth0/queues/tx-1/xps_cpus
Writing '4096' to /proc/sys/net/core/somaxconn
Writing '4096' to /proc/sys/net/ipv4/tcp_max_syn_backlog
```

Dataset: 1.500.000.000 records, 5 columns, fixed 64B size
Loaders: 4 _c4.2xlarge_ instances running cassadra-stress with the following parameters:
`cassandra-stress read no-warmup duration=1h cl=ONE -mode native cql3 connectionsPerHost=90 -col n=FIXED\(5\) size=FIXED\(64\) -pop dist=GAUSSIAN\(1..150000000,75000000,25000000\) -node <node-IP> -rate threads=200 limit=100000/s -log file=<logfile> -schema "replication(factor=1)" -errors ignore`

SoftIRQ using 100% on cpu0 and none in cpu16. Also hurting throughput since we can not get passed 75% load on the AMI. Evidence here (see CPU SoftIRQ under the Network section) and here (prometheus data).

==========================

Solution: 

Install newer driver version.

```
sudo yum upgrade
sudo rpm -ih --nodeps kmod-ena-1.5.3-1.el7.centos.7.x86_64.rpm
```

```
ethtool -i eth0
driver: *ena*
version: *1.5.3g*
firmware-version:
expansion-rom-version:
bus-info: 0000:00:03.0
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: no
```

**Comment by gnumoreno** [comment]
@vladzcloudius, please add any details I may have missed.

**Comment by syuu1228** [comment]
We have fix for this on scylla-ami, but it's not merged into branch-2.1 (available on branch-2.2 and master). Need to backport:
https://github.com/scylladb/scylla-ami/commit/5170011fbc478a93d9f03d12b214f7df4c678b91#diff-6ea74efa9c4f13dcc2c2e52e1a8f5180

**Comment by avikivity** [comment]
master uses kernel-ml, will the kmod work on the centos kernel?

**Comment by syuu1228** [comment]
@avikivity Since it's kernel module source code + dkms, it should work on multiple kernel versions unlike binary module, and also we had built ixgbevf / ena drivers for CentOS kernel on older AMIs, I suppose it's works finely.
But if we want to do that we need to modify the commit not to use kernel-ml, I'll send new one for branch-2.1

**Comment by syuu1228** [comment]
I posted patchset to install recent ixgbevf/ENA driver on AMI that backported from master, but it doesn't fix the problem if users keep using 2.1.x AMI instances, even they are keep using latest Scylla by "yum update", since we don't provide ixgbevf/ENA drivers.
I also think it's not just NIC driver problem, we don't have any mechanism to upgrade scylla-ami scripts on AMI instances.
Even users upgraded latest Scylla, old scylla-ami scripts remains.

**Comment by slivne** [comment]
in 2.2 (49896ec5d183532b9e40546a4c0b58b7d0c1d3e6) / master this is fixed (3567f0828bc987be2370c491d2f2f61fb7e74198)

2.1 backport still not merged

@syuu1228 - am I missing something ?

**Comment by tzach** [comment]
merged to 2.2, closing this issue
@avikivity note this was not port to 2.1
